ThirdParty/ecec/src/encrypt.c

```diff
@@ -0,0 +1,292 @@
+  if (!EC_POINT_point2oct(
+        EC_KEY_get0_group(senderPrivKey), EC_KEY_get0_public_key(senderPrivKey),
+        POINT_CONVERSION_UNCOMPRESSED, &payload[ECE_AES128GCM_HEADER_LENGTH],
+        ECE_WEBPUSH_PUBLIC_KEY_LENGTH, NULL)) {
+    err = ECE_ERROR_ENCODE_PUBLIC_KEY;
+    goto end;
+  }
+
+  bool isLastRecord = false;
+  size_t overhead = ECE_AES128GCM_MIN_RS - 1;
+  size_t plaintextPerBlock = rs - overhead;
+  size_t blockStart = 0;
+  size_t recordStart =
+    ECE_AES128GCM_HEADER_LENGTH + ECE_WEBPUSH_PUBLIC_KEY_LENGTH;
+  size_t counter = 0;
+  while (!isLastRecord) {
+    // Pad so that at least one plaintext byte is in a block.
+    size_t blockPadLen = plaintextPerBlock - 1;
+    if (blockPadLen > padLen) {
+      blockPadLen = padLen;
+    }
+    if (padLen && !blockPadLen) {
+      // If our record size is `ECE_AES128GCM_MIN_RS + 1`, we can only include
+      // one byte of data, so write the padding first.
+      blockPadLen++;
+    }
+    padLen -= blockPadLen;
+
+    size_t blockEnd = blockStart + plaintextPerBlock - blockPadLen;
+    if (blockEnd >= plaintextLen) {
+      blockEnd = plaintextLen;
+      if (!padLen) {
+        // We've reached the last record when the plaintext and padding are
+        // exhausted.
+        isLastRecord = true;
+      }
+    }
+    size_t blockLen = blockEnd - blockStart;
+
+    pad[0] = isLastRecord ? 2 : 1;
+
+    size_t recordEnd = recordStart + blockLen + blockPadLen + overhead;
+    if (recordEnd >= maxRecordEnd) {
+      recordEnd = maxRecordEnd;
+    }
+
+    // Generate the IV for this record using the nonce.
+    uint8_t iv[ECE_NONCE_LENGTH];
+    ece_generate_iv(nonce, counter, iv);
+
+    // Encrypt and pad the block. `blockPadLen + 1` ensures we always write the
+    // delimiter.
+    err = ece_aes128gcm_encrypt_block(ctx, key, iv, &plaintext[blockStart],
+                                      blockEnd - blockStart, pad,
+                                      blockPadLen + 1, &payload[recordStart]);
+    if (err) {
+      goto end;
+    }
+    blockStart = blockEnd;
+    recordStart = recordEnd;
+    counter++;
+  }
+  *payloadLen = recordStart;
+
+end:
+  EVP_CIPHER_CTX_free(ctx);
+  free(pad);
+  return err;
+}
+
+size_t
+ece_aes128gcm_payload_max_length(uint32_t rs, size_t padLen,
+                                 size_t plaintextLen) {
+  if (rs < ECE_AES128GCM_MIN_RS) {
+    return 0;
+  }
+  // The per-record overhead for the padding delimiter and authentication tag.
+  size_t overhead = ECE_AES128GCM_MIN_RS - 1;
+  // The total length of the data to encrypt, including the plaintext and
+  // padding.
+  size_t dataLen = plaintextLen + padLen;
+  // The maximum length of data to include in each record, excluding the
+  // padding delimiter and authentication tag.
+  size_t dataPerBlock = rs - overhead;
+  // The total number of encrypted records.
+  size_t numRecords = (dataLen / dataPerBlock) + 1;
+  return ECE_AES128GCM_HEADER_LENGTH + ECE_AES128GCM_MAX_KEY_ID_LENGTH +
```

Comment:
As you noted in https://github.com/web-push-libs/ecec/issues/20 the result can overflow. But not only that, intermediate values (as described above) can do as well.